lib/pbench/cli/server/shell.py

```diff
@@ -40,42 +42,135 @@ def find_the_unicorn(logger: Logger):
+class PbenchIndexer:
+    """Managed the context of the indexer subprocesses."""
 
-    If a crontab file already exists, no action is taken, otherwise a crontab
-    file is created using `pbench-create-crontab` and then installed using the
-    `crontab` command.
+    def __init__(self, bin_dir: Path, cwd: Path, logger: Logger):
+        """Construct the subprocess tracker state.
 
-    If either of those operations fail, the crontab file is removed.
+        Args:
+            bin_dir: Where to find the `pbench-index` command
+            cwd: Where to set the command working directory
+            logger: Python logger for feedback.
+        """
+        self.bin_dir = bin_dir
+        self.cwd = cwd
+        self.logger = logger
+        self.indexer: Optional[subprocess.Popen] = None
+        self.reindexer: Optional[subprocess.Popen] = None
+        self.shutoff: Event = Event()
+        self.reaper: Optional[Thread] = None
 
-    Return 0 on success, 1 on failure.
-    """
-    ret_val = 0
-    crontab_f = Path(crontab_dir) / "crontab"
-    if not crontab_f.exists():
-        os.environ["PATH"] = ":".join([str(bin_dir), os.environ["PATH"]])
-        # Create the crontab file from the server configuration.
-        cp = subprocess.run(["pbench-create-crontab", crontab_dir], cwd=cwd)
-        if cp.returncode != 0:
-            logger.error(
-                "Failed to create crontab file from configuration: {}", cp.returncode
-            )
-            ret_val = 1
-        else:
-            # Install the created crontab file.
-            cp = subprocess.run(["crontab", f"{crontab_dir}/crontab"], cwd=cwd)
-            if cp.returncode != 0:
-                logger.error("Failed to install crontab file: {}", cp.returncode)
-                ret_val = 1
-    if ret_val != 0:
-        crontab_f.unlink(missing_ok=True)
-    return ret_val
-
-
-def run_gunicorn(server_config: PbenchServerConfig, logger: Logger) -> int:
+    def start_indexer(self):
+        """Isolate starting the primary indexer as a subprocess.
+
+        NOTE: we specify the 'python3' command explicitly rather than relying
+        on the shebang to avoid the intermediate `env python` command which
+        complicates reaping.
+        """
+        self.indexer = subprocess.Popen(
+            ["python3", self.bin_dir / "pbench-index"], cwd=self.cwd
+        )
+        self.logger.info("Index process running in pid {}", self.indexer.pid)
+
+    def start_reindexer(self):
+        """Isolate starting the secondary reindexer as a subprocess.
+
+        TODO: I'm not sure we really want to retain the old "reindex". For one
+        thing, re-indexing with existing index documents may be problematic,
+        especially if the new indexer doesn't overwrite every document. A
+        better mechanism would be to delete all the existing documents using
+        the index-map and then we can simply READY the INDEX operation.
```

Comment:
> Unfortunately, we since we keep more than one set of result tar ball JSON documents in one index.

@portante, I think there was a glitch at a key moment in that statement....  (Did we lose the word "can't"?  If so, can't what?...delete the indices directly?)

Reply:
Updated, thanks for catching that.

Reply:
I expect that re-indexing will be a rare operation.  It strikes me that there are two principal motivations for it:
- a particular result's index was "damaged" (or incomplete or otherwise unsatisfactory), in which case it should be fine to delete all the documents by result-ID before indexing it again; or,
- we're changing the index map, in which case we'll presumably blow away the entire index and re-index all the results which belong in it.

Either way, we don't really need/want a "reindex" operation -- we want to be able to delete either a single result or a whole index and then index the particular result or a list of 'em.

Reply:
Again, we can easily remove every ES document based on the `index-map`. If we want a `pbench-reindex` it could eliminate the existing documents for each dataset and then trigger a new `INDEX` operation. And, yeah, if we're changing the indexer we might want to simply delete all the indices and start from scratch; which is really a completely different operation. (We'll need something to just delete all `index-map` metadata and trigger `INDEX` for everything.)

But it sounds like I was right in my initial thought that we just shouldn't bother with trying to retain the concept of `REINDEX`, so we can drop the subprocess entirely. Cool.